overlord/patch/patch4.go

```diff
@@ -0,0 +1,261 @@
+	var ss patch4SnapSetup
+
+	switch err := p4.get(task, "snap-setup", true, &ss); err {
+	case state.ErrNoState:
+		// continue below
+	case nil:
+		return &ss, nil
+	default:
+		return nil, err
+	}
+
+	var id string
+	if err := p4.get(task, "snap-setup-task", false, &id); err != nil {
+		return nil, err
+	}
+
+	if err := p4.get(task.State().Task(id), "snap-setup", false, &ss); err != nil {
+		return nil, err
+	}
+
+	return &ss, nil
+}
+
+func (p4 patch4T) snapSetupAndState(task *state.Task) (*patch4SnapSetup, *patch4SnapState, error) {
+	var snapst patch4SnapState
+
+	ss, err := p4.taskSnapSetup(task)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var snaps map[string]*json.RawMessage
+	err = task.State().Get("snaps", &snaps)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot get snaps state: %v", err)
+	}
+	raw, ok := snaps[ss.Name()]
+	if !ok {
+		return nil, nil, fmt.Errorf("cannot get snap state for %q: %v", ss.Name(), err)
+	}
+	err = json.Unmarshal([]byte(*raw), &snapst)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot get state for snap %q: %v", ss.Name(), err)
+	}
+
+	return ss, &snapst, err
+}
+
+func (patch4T) get(task *state.Task, key string, passThroughMissing bool, value interface{}) error {
+	err := task.Get(key, value)
+	if err == nil || (passThroughMissing && err == state.ErrNoState) {
+		return err
+	}
+	change := task.Change()
+
+	return fmt.Errorf("cannot get %q from task %s (%s) of change %s (%s): %v",
+		key, task.ID(), task.Kind(), change.ID(), change.Kind(), err)
+}
+
+func (p4 patch4T) addCleanup(task *state.Task) error {
+	// NOTE we could check for the status of the change itself, but
+	// copy-snap-data is the one creating the trash, so if it's run there's
+	// no sense in fiddling with the change.
+	if task.Status().Ready() {
+		return nil
+	}
+
+	ss, err := p4.taskSnapSetup(task)
+	if err != nil {
+		return err
+	}
+
+	var tid string
+	if err := p4.get(task, "snap-setup-task", false, &tid); err != nil {
+		return err
+	}
+
+	change := task.Change()
+	revisionStr := ""
+	if ss.SideInfo != nil {
+		revisionStr = fmt.Sprintf(" (%s)", ss.Revision())
+	}
+
+	tasks := change.Tasks()
+	last := tasks[len(tasks)-1]
+	newTask := task.State().NewTask("cleanup", fmt.Sprintf("Clean up %q%s install", ss.Name(), revisionStr))
+	newTask.Set("snap-setup-task", tid)
+	newTask.WaitFor(last)
+	change.AddTask(newTask)
+
+	return nil
+}
+
+func (p4 patch4T) mangle(task *state.Task) error {
+	ss, snapst, err := p4.snapSetupAndState(task)
+	if err != nil {
+		return err
+	}
+
+	var hadCandidate bool
+	if err := p4.get(task, "had-candidate", false, &hadCandidate); err != nil {
+		return err
+	}
+
+	if hadCandidate {
+		change := task.Change()
+		if change.Kind() != "revert-snap" {
+			return fmt.Errorf("had-candidate true for task %s (%s) of non-revert change %s (%s)",
+				task.ID(), task.Kind(), change.ID(), change.Kind())
+		}
+	}
+
+	task.Clear("had-candidate")
```

Comment:
no; that would leave leftover `had-candidate` keys in the state (with false values). If the state didn't have the key the method would've exited earlier.